File: Cpp/src/Order.cpp

#include "Order.h"
#include "Product.h"



double Order::order_total_dollars()
{
    double dollars = 0.0;
    for (int i = 0; i < this->_products.size(); i++)
        dollars += _products[i]->get_product_price()->get_price_in_currency("USD");
    return dollars;
}

void Order::save_order_to_database()
{
    printf("Unsupported Operation %s\n",
           "missing from this exercise - shouldn't be called from a unit test");
    exit(1);
}

std::string Order::get_order_id() {
    return _id;
}

std::vector<Product *> Order::get_order_products() {
    return _products;
}

time_t Order::get_order_date() {
    return _date;
}

void Order::add_order_products(std::vector<Product *> products) {
    for (int i = 0; i < products.size(); ++i) {
        _products.push_back(products[i]);
    }
}

File: TypeScript/src/Order.ts
import { Product } from "./Product";
import { Store } from "./Store";

export class Order {

    private readonly id: string;
    private readonly date: Date;
    private products: Product[];
    private store: Store;

    constructor(id: string, date: Date, store: Store, products: Product[]) {
        this.id = id;
        this.date = date;
        this.products = [];
        this.store = store;
        this.addProducts(...products);
    }

    totalDollars(): number {
        return this.products.reduce((total, product) => total + product.getPrice().getAmountInCurrency("USD"), 0);
    }

    toString(): string {
        return "Order{" + this.id + "}";
    }

    saveToDatabase(): void {
        throw new Error("missing from this exercise - shouldn't be called from a unit test");
    }

    getId(): string {
        return this.id;
    }

    getProducts(): Product[] {
        return this.products;
    }

    getDate(): Date {
        return this.date;
    }

    addProducts(...products: Product[]) {
        this.products.push(...products);
    }

    getStore(): Store {
        return this.store;
    }

}
